ts/nni_manager/common/trial_keeper/keeper.ts

```diff
@@ -0,0 +1,154 @@
+import path from 'path';
+
+import type { Command } from 'common/command_channel/interface';
+import { HttpChannelServer } from 'common/command_channel/http';
+import globals from 'common/globals';
+import { Logger, getLogger } from 'common/log';
+import { TrialProcess, TrialProcessOptions } from './process';
+import { GpuSchedulerClient } from './gpu_scheduler_client';
+
+export declare namespace TrialKeeper {
+    export interface TrialOptions {
+        id: string;
+        command: string;
+        codeDirectoryName: string;
+        sequenceId?: number;
+        gpuNumber?: number;
+        gpuRestrictions?: GpuRestrictions;
+    }
+
+    export interface GpuRestrictions {
+        onlyUseIndices?: number[];
+        rejectActive?: boolean;
+        rejectComputeActive?: boolean;
+    }
+}
+
+export class TrialKeeper {
+    private envId: string;
+    private channels: HttpChannelServer;
+    private dirs: Map<string, string> = new Map();
+    private emitter: EventEmitter = new EventEmitter();
+    private gpuScheduler: GpuSchedulerClient;
+    private log: Logger;
+    private platform: string;
+    private trials: Map<string, TrialProcess> = new Map();
+
+    constructor(environmentId: string, platform: string, enableGpuScheduler: boolean) {
+        this.envId = environmentId;
+        this.platform = platform;
+        this.log = getLogger(`TrialKeeper.${environmentId}`);
+
+        this.gpuScheduler = new GpuSchedulerClient(enableGpuScheduler);
+
+        this.channels = new HttpChannelServer(this.envId, `/env/${this.envId}`);
+        this.channels.onReceive((trialId, command) => {
+            this.emitter.emit('command', trialId, command);
+            if (command.type !== 'request_parameter' && command.type !== 'metric') {
+                this.log.warning(`Unexpected command from trial ${trialId}:`, command);
+            }
+        });
+    }
+
+    // TODO: support user configurable init command
+    public async start(): Promise<void> {
+        await Promise.all([
+            this.gpuScheduler.start(),
+            this.channels.start(),
+        ]);
+    }
+
+    public async shutdown(): Promise<void> {
+        let promises: Promise<void>[] = [
+            this.gpuScheduler.shutdown(),
+            this.channels.shutdown(),
+        ];
+
+        const trials = Array.from(this.trials.values());
+        promises = promises.concat(trials.map(trial => trial.kill()));
+
+        await Promise.all(promises);
+    }
+
+    public registerDirectory(name: string, path: string): void {
+        this.dirs.set(name, path);
+    }
+
+    // FIXME: the method name will be changed when we support distributed trials
+    public async createTrial(options: TrialKeeper.TrialOptions): Promise<boolean> {
+        const trialId = options.id;
+
+        const gpuEnv = await this.gpuScheduler.schedule(trialId, options.gpuNumber, options.gpuRestrictions);
+        if (gpuEnv === null) {
+            this.log.info('No GPU available');
```

Comment:
Maybe this should be "Scheduling failed because the required constraint ${sku} can't be satisfied."